plutus-benchmark/flat/Report.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Report ( Size(..)
+              , Measures
+              , measureSize
+              , withRatio
+              , showContractMeasures
+              , showCodecSizes
+              ) where
+
+import qualified Data.ByteString              as BS
+import           Data.List                    (sort)
+import           Data.Map                     (Map)
+import qualified Data.Map                     as Map
+import           Data.Text                    (Text, pack, unpack)
+import           GHC.Generics
+import           Text.PrettyPrint.Boxes
+
+import           Language.PlutusCore          (Name (..))
+import           Language.PlutusCore.Universe
+import           Language.UntypedPlutusCore
+import           Prelude                      as P
+
+type Tm = Term Name DefaultUni ()
+
+data Size = Size
+  { szContract :: Text
+  , szCodec    :: Text
+  , szBytes    :: Integer
+  } deriving (Eq, Show, Generic)
+
+type Measures = Map Text [Size]
+
+-- | Order by contract name and serialized contract size.
+instance Ord Size where
+  compare Size {szContract=leftContract , szBytes=leftSize }
+          Size {szContract=rightContract, szBytes=rightSize} =
+    compare leftContract rightContract P.<> compare leftSize rightSize
+
+measureSize ::
+     Map (Text, Text) BS.ByteString
+  -> Map Text [Size]
+measureSize =
+  Map.map sort .
+  Map.foldlWithKey go Map.empty
+  where
+    go :: Map Text [Size] -> (Text, Text) -> BS.ByteString -> Map Text [Size]
+    go ms (contractName, codecName) contract =
+      let oldMs = Map.findWithDefault [] contractName ms
+          sizeMs = Size { szContract = contractName
+                        , szCodec    = codecName
+                        , szBytes    = fromIntegral $ BS.length contract
+                        }
+          newMs = sizeMs : oldMs
+      in Map.insert contractName newMs ms
+
+withRatio :: Map Text [Size]
+          -> Map Text [(Size, Float)]
+withRatio =
+  Map.map (go 0)
+  where
+    go :: Float -> [Size] -> [(Size, Float)]
+    go _  []        = []
+    go 0  (hd : tl) = (hd, 1) : go (fromInteger $ szBytes hd) tl
+    go bs (hd : tl) = (hd, fromInteger (szBytes hd) / bs) : go bs tl
+
+showContractMeasures :: Map Text [(Size, Float)]
+              -> Text
+showContractMeasures =
+  Map.foldlWithKey go ""
```

Comment:
I bet this would come out neater if you used `Map.foldMapWithKey` and relied on the `Monoid` instance for strings.

Reply:
Indeed, it did.